Author: Yvan Roux <yvan.roux@linaro.org>
Description: Fix PR target/57909.
 * config/arm/arm.c (gen_movmem_ldrd_strd):
   Fix unaligned load/store usage in HI mode.

--- trunk/gcc/config/arm/arm.c	2013/07/17 11:15:24	201004
+++ trunk/gcc/config/arm/arm.c	2013/07/17 11:39:14	201005
@@ -12018,8 +12018,16 @@
       dst = adjust_address (dst, HImode, 0);
       src = adjust_address (src, HImode, 0);
       reg0 = gen_reg_rtx (SImode);
-      emit_insn (gen_unaligned_loadhiu (reg0, src));
-      emit_insn (gen_unaligned_storehi (dst, gen_lowpart (HImode, reg0)));
+      if (src_aligned)
+        emit_insn (gen_zero_extendhisi2 (reg0, src));
+      else
+        emit_insn (gen_unaligned_loadhiu (reg0, src));
+
+      if (dst_aligned)
+        emit_insn (gen_movhi (dst, gen_lowpart(HImode, reg0)));
+      else
+        emit_insn (gen_unaligned_storehi (dst, gen_lowpart (HImode, reg0)));
+
       src = next_consecutive_mem (src);
       dst = next_consecutive_mem (dst);
       if (len == 2)
